catalyst-gateway/bin/src/event_db/cardano/cip36_registration/mod.rs

```diff
@@ -33,95 +38,149 @@ const PUBLIC_VOTING_KEY_COLUMN: &str = "public_voting_key";
+    pub stats: serde_json::Value,
+    /// Whether the registration is valid.
+    pub valid: bool,
+}
 
-        Ok(())
+impl IndexedVoterRegistrationParams {
+    /// Creates voter registration indexing data from block data.
+    pub(crate) fn from_block_data(
+        block: &MultiEraBlock, network: Network,
+    ) -> Option<Vec<IndexedVoterRegistrationParams>> {
+        if !valid_era(block.era()) {
+            return None;
+        }
+
+        let registrations = block
+            .txs()
+            .into_iter()
+            .filter_map(|tx| {
+                if tx.metadata().is_empty() {
+                    return None;
+                }
+
+                let Some(cip36_metadata) =
+                    Cip36Metadata::generate_from_tx_metadata(&tx.metadata(), network)
+                else {
+                    return None;
+                };
+
+                let (stake_credential, voting_info, rewards_address, nonce) =
+                    if let Some(reg) = cip36_metadata.registration {
+                        (
+                            Some(reg.stake_key.get_credentials().to_vec()),
+                            Some(reg.voting_info),
+                            Some(reg.rewards_address.0),
+                            Some(reg.nonce.0),
+                        )
+                    } else {
+                        (None, None, None, None)
+                    };
+
+                let encoded_voting_key = if let Some(voting_info) = voting_info.as_ref() {
+                    let Ok(enc) = serde_json::to_string(voting_info) else {
+                        return None;
+                    };
+
+                    Some(enc.into_bytes())
+                } else {
+                    None
+                };
+
+                let multiple_delegations = voting_info.as_ref().is_some_and(|vi| {
+                    if let VotingInfo::Delegated(delegations) = vi {
+                        delegations.len() > 1
+                    } else {
+                        false
+                    }
+                });
+
+                let is_valid = !multiple_delegations
+                    && stake_credential.is_some()
+                    && encoded_voting_key.is_some()
+                    && rewards_address.is_some()
+                    && cip36_metadata.raw_metadata.is_some()
+                    && nonce.is_some()
+                    && cip36_metadata.errors_report.is_empty();
+
+                Some(IndexedVoterRegistrationParams {
+                    tx_id: tx.hash().to_vec(),
+                    stake_credential,
+                    public_voting_key: encoded_voting_key,
+                    payment_address: rewards_address,
+                    nonce,
+                    cip36_metadata: cip36_metadata.raw_metadata,
+                    stats: serde_json::json!(cip36_metadata.errors_report),
+                    valid: is_valid,
+                })
+            })
+            .collect();
```

Comment:
Is it possible to break down this logic into multiple functions?

Reply:
I think that should go in another PR since this was already implemented and I just moved it.